### ISSUE TITLE
[Study] vSOME/IP

### ISSUE THREAD

**Issue by LWJ0513** [FALCON-RACER/DES_Head-Unit#1]

**Comment by tweetyboo** [comment]
Server-side
`#include <vsomeip/vsomeip.hpp>
#include <iostream>
#include <chrono>
#include <thread>

class VehicleSpeedPublisher {
public:
    VehicleSpeedPublisher() : app_(vsomeip::runtime::get()->create_application("SpeedService")) {}

    void init() {
        app_->init();
        app_->register_message_handler(SERVICE_ID, INSTANCE_ID, EVENT_ID, 
                                       std::bind(&VehicleSpeedPublisher::on_get_speed, this, std::placeholders::_1));
        
        app_->offer_service(SERVICE_ID, INSTANCE_ID);
        
        // Register event group and event ID for notifications
        app_->offer_event(SERVICE_ID, INSTANCE_ID, EVENT_ID, EVENTGROUP_ID);

        std::thread([this]() {
            // Periodically notify subscribers with speed updates
            while (true) {
                notify_subscribers();
                std::this_thread::sleep_for(std::chrono::seconds(5));  // Notify every 5 seconds
            }
        }).detach();

        app_->start();
    }

private:
    std::shared_ptr<vsomeip::application> app_;
    int current_speed_ = 0;

    void notify_subscribers() {
        current_speed_ = (current_speed_ + 10) % 120; // Simulate speed changes
        std::shared_ptr<vsomeip::message> msg = vsomeip::runtime::get()->create_notification(SERVICE_ID, INSTANCE_ID, EVENT_ID);
        
        vsomeip::payload_ptr_t payload = vsomeip::runtime::get()->create_payload();
        std::vector<vsomeip::byte_t> data = { static_cast<vsomeip::byte_t>(current_speed_) };
        payload->set_data(data);
        msg->set_payload(payload);
        
        std::cout << "Notifying subscribers: Speed = " << current_speed_ << " km/h" << std::endl;
        app_->notify(msg);
    }

    void on_get_speed(const std::shared_ptr<vsomeip::message> &request) {
        // Handle requests (if applicable) for current speed from clients
        std::cout << "Received speed request." << std::endl;
    }
};

// Service and event IDs
#define SERVICE_ID 0x1234
#define INSTANCE_ID 0x5678
#define EVENT_ID 0x4321
#define EVENTGROUP_ID 0x8765

int main() {
    VehicleSpeedPublisher speedService;
    speedService.init();
    return 0;
}
`

**Comment by tweetyboo** [comment]
Client-Side
`#include <vsomeip/vsomeip.hpp>
#include <iostream>
#include <thread>

class VehicleSpeedSubscriber {
public:
    VehicleSpeedSubscriber() : app_(vsomeip::runtime::get()->create_application("SpeedClient")) {}

    void init() {
        app_->init();

        app_->register_message_handler(SERVICE_ID, INSTANCE_ID, EVENT_ID,
                                       std::bind(&VehicleSpeedSubscriber::on_speed_update, this, std::placeholders::_1));
        
        app_->request_service(SERVICE_ID, INSTANCE_ID);
        app_->subscribe(SERVICE_ID, INSTANCE_ID, EVENTGROUP_ID, EVENT_ID);

        app_->start();
    }

private:
    std::shared_ptr<vsomeip::application> app_;

    void on_speed_update(const std::shared_ptr<vsomeip::message> &msg) {
        auto payload = msg->get_payload();
        std::vector<vsomeip::byte_t> data(payload->get_data(), payload->get_data() + payload->get_length());
        int speed = data[0]; // Assume that speed is sent as a single byte

        std::cout << "Received speed update: " << speed << " km/h" << std::endl;
    }
};

// Service and event IDs
#define SERVICE_ID 0x1234
#define INSTANCE_ID 0x5678
#define EVENT_ID 0x4321
#define EVENTGROUP_ID 0x8765

int main() {
    VehicleSpeedSubscriber speedClient;
    speedClient.init();
    return 0;
}
`

**Comment by tweetyboo** [comment]
### Server-Side Example for VSOME/IP Subscribe/Notify Mechanism

```cpp
#include <vsomeip/vsomeip.hpp>
#include <iostream>
#include <chrono>
#include <thread>

class VehicleSpeedPublisher {
public:
    VehicleSpeedPublisher() : app_(vsomeip::runtime::get()->create_application("SpeedService")) {}

    void init() {
        app_->init();
        app_->register_message_handler(SERVICE_ID, INSTANCE_ID, EVENT_ID, 
                                       std::bind(&VehicleSpeedPublisher::on_get_speed, this, std::placeholders::_1));
        
        app_->offer_service(SERVICE_ID, INSTANCE_ID);
        
        // Register event group and event ID for notifications
        app_->offer_event(SERVICE_ID, INSTANCE_ID, EVENT_ID, EVENTGROUP_ID);

        std::thread([this]() {
            // Periodically notify subscribers with speed updates
            while (true) {
                notify_subscribers();
                std::this_thread::sleep_for(std::chrono::seconds(5));  // Notify every 5 seconds
            }
        }).detach();

        app_->start();
    }

private:
    std::shared_ptr<vsomeip::application> app_;
    int current_speed_ = 0;

    void notify_subscribers() {
        current_speed_ = (current_speed_ + 10) % 120; // Simulate speed changes
        std::shared_ptr<vsomeip::message> msg = vsomeip::runtime::get()->create_notification(SERVICE_ID, INSTANCE_ID, EVENT_ID);
        
        vsomeip::payload_ptr_t payload = vsomeip::runtime::get()->create_payload();
        std::vector<vsomeip::byte_t> data = { static_cast<vsomeip::byte_t>(current_speed_) };
        payload->set_data(data);
        msg->set_payload(payload);
        
        std::cout << "Notifying subscribers: Speed = " << current_speed_ << " km/h" << std::endl;
        app_->notify(msg);
    }

    void on_get_speed(const std::shared_ptr<vsomeip::message> &request) {
        // Handle requests (if applicable) for current speed from clients
        std::cout << "Received speed request." << std::endl;
    }
};

// Service and event IDs
#define SERVICE_ID 0x1234
#define INSTANCE_ID 0x5678
#define EVENT_ID 0x4321
#define EVENTGROUP_ID 0x8765

int main() {
    VehicleSpeedPublisher speedService;
    speedService.init();
    return 0;
}
```

**Comment by tweetyboo** [comment]
### Client-Side Example for VSOME/IP Subscribe/Notify Mechanism

```cpp
#include <vsomeip/vsomeip.hpp>
#include <iostream>
#include <thread>

class VehicleSpeedSubscriber {
public:
    VehicleSpeedSubscriber() : app_(vsomeip::runtime::get()->create_application("SpeedClient")) {}

    void init() {
        app_->init();

        app_->register_message_handler(SERVICE_ID, INSTANCE_ID, EVENT_ID,
                                       std::bind(&VehicleSpeedSubscriber::on_speed_update, this, std::placeholders::_1));
        
        app_->request_service(SERVICE_ID, INSTANCE_ID);
        app_->subscribe(SERVICE_ID, INSTANCE_ID, EVENTGROUP_ID, EVENT_ID);

        app_->start();
    }

private:
    std::shared_ptr<vsomeip::application> app_;

    void on_speed_update(const std::shared_ptr<vsomeip::message> &msg) {
        auto payload = msg->get_payload();
        std::vector<vsomeip::byte_t> data(payload->get_data(), payload->get_data() + payload->get_length());
        int speed = data[0]; // Assume that speed is sent as a single byte

        std::cout << "Received speed update: " << speed << " km/h" << std::endl;
    }
};

// Service and event IDs
#define SERVICE_ID 0x1234
#define INSTANCE_ID 0x5678
#define EVENT_ID 0x4321
#define EVENTGROUP_ID 0x8765

int main() {
    VehicleSpeedSubscriber speedClient;
    speedClient.init();
    return 0;
}
```
